source/solvers/heat_transfer_assemblers.cc

```diff
@@ -0,0 +1,607 @@
+                                     physical_properties.fluids[1].viscosity);
+
+          specific_heat = calculate_point_property(
+            scratch_data.phase_values[q],
+            physical_properties.fluids[0].specific_heat,
+            physical_properties.fluids[1].specific_heat);
+
+          thermal_conductivity = calculate_point_property(
+            scratch_data.phase_values[q],
+            physical_properties.fluids[0].thermal_conductivity,
+            physical_properties.fluids[1].thermal_conductivity);
+
+          // Useful definitions
+          dynamic_viscosity = viscosity * density;
+          rho_cp            = density * specific_heat;
+          alpha             = thermal_conductivity / rho_cp;
+        }
+
+      // Store JxW in local variable for faster access
+      const double JxW = scratch_data.fe_values_T.JxW(q);
+
+      const auto velocity          = scratch_data.velocity_values[q];
+      const auto velocity_gradient = scratch_data.velocity_gradient_values[q];
+
+      // Calculation of the magnitude of the velocity for the
+      // stabilization parameter
+      const double u_mag = std::max(velocity.norm(), 1e-12);
+
+      // Calculation of the GLS stabilization parameter. The
+      // stabilization parameter used is different if the simulation is
+      // steady or unsteady. In the unsteady case it includes the value
+      // of the time-step
+      const double tau =
+        is_steady(method) ?
+          1. / std::sqrt(std::pow(2. * rho_cp * u_mag / h, 2) +
+                         9 * std::pow(4 * alpha / (h * h), 2)) :
+          1. /
+            std::sqrt(std::pow(sdt, 2) + std::pow(2. * rho_cp * u_mag / h, 2) +
+                      9 * std::pow(4 * alpha / (h * h), 2));
+
+      // Calculate the strong residual for GLS stabilization
+      strong_residual_vec[q] += rho_cp * velocity * temperature_gradient -
+                                thermal_conductivity * temperature_laplacian;
+
+
+      for (unsigned int i = 0; i < n_dofs; ++i)
+        {
+          const auto phi_T_i      = scratch_data.phi_T[q][i];
+          const auto grad_phi_T_i = scratch_data.grad_phi_T[q][i];
+
+          // rhs for : - k * laplacian T + rho * cp * u * grad T - f
+          // -grad(u)*grad(u) = 0
+          local_rhs(i) -=
+            (thermal_conductivity * grad_phi_T_i * temperature_gradient +
+             rho_cp * phi_T_i * velocity * temperature_gradient -
+             scratch_data.source[q] * phi_T_i) *
+            JxW;
+
+          if (this->simulation_parameters.multiphysics.viscous_dissipation)
+            {
+              local_rhs(i) -= (-dynamic_viscosity * phi_T_i *
+                               scalar_product(velocity_gradient +
+                                                transpose(velocity_gradient),
+                                              transpose(velocity_gradient))) *
+                              JxW;
+            }
```

Comment:
I would put this into another assembler that just assembles an RHS instead of putting an if there.